helix-term/src/compositor.rs

```diff
@@ -138,16 +139,22 @@ impl Compositor {
 
         // propagate events through the layers until we either find a layer that consumes it or we
         // run out of layers (event bubbling)
-        for layer in self.layers.iter_mut().rev() {
```

Comment:
this takes a mutual reference to `self`, since the match clause of  `EventResult::Used(...)` does not do an early return, borrow checker will complain

hopefully, soon we can have a mutable streaming iterator when GATs are stabilized https://github.com/rust-lang/rust/issues/44265
but for now, using an indexed for loop seems the easiest

Reply:
The borrow checker is highlighting an issue: you shouldn't mutate `self`.`self.layers` while iterating over the layers. For example a callback could remove half of the layer stack, then your next iteration would panic on `self.layers[i].as_mut()`.

Reply:
We should probably collect all the callbacks and then run them in order.

Reply:
Ah you are right, the whole compositor gets passed to each callback at mutable ref, so it's layers may be removed, invalidating the index range, while iterating over them...